### DISCUSSION TITLE
asyncio uart question

### DISCUSSION THREAD

**Discussion by vahithosan** [micropython/micropython#12840] [ESP32]
Hi.
I want to receive data over Uart.  I tried with asyncio. 

```
MicroPython 95d8b5f-dirty on 2023-10-31; Generic ESP32 module with ESP32
Type "help()" for more information.
>>> 
paste mode; Ctrl-C to cancel, Ctrl-D to finish

=== import uasyncio as asyncio
=== from machine import UART
=== uart1 = UART(2, tx=22, rx=21)
=== uart1.init(baudrate=9600, bits=8, parity=0, stop=2)
=== 
=== async def receiver():
===     sreader = asyncio.StreamReader(uart1)
===     while True:
===         res = await sreader.readline()
===         print('Recieved', res)
=== 
=== loop = asyncio.get_event_loop()
=== loop.create_task(receiver())
=== loop.run_forever()
```

I got an error.  

```
Task exception wasn't retrieved
future: <Task> coro= <generator object 'receiver' at 3ffd8420>
Traceback (most recent call last):
  File "asyncio/core.py", line 1, in run_until_complete
  File "<stdin>", line 9, in receiver
  File "asyncio/stream.py", line 1, in readline
TypeError: unsupported types for __add__: 'bytes', 'NoneType'

```
When I try without asyncio

```
=== from machine import UART
=== 
=== uart = UART(2, tx=22, rx=21)
=== uart.init(baudrate=9600, bits=8, parity=0, stop=2)

```

```
>>> uart.readline()  
b'T 022\r'

```
No problem. 

What should I do for asyncio?

**Comment by jimmo** [comment]
@vahithosan 

> `MicroPython 95d8b5f-dirty on 2023-10-31;`

Just out of curiosity, where did you get this firmware from? Looks like something has gone wrong with the version information string.

> `TypeError: unsupported types for __add__: 'bytes', 'NoneType'`

This suggests that the uart is returning `None` from `readline()`. Which it's not supposed to do because the StreamReader shouldn't have been woken up unless the uart was readable.

I will need to try and replicate this here.

**Comment by vahithosan** [comment]
> Just out of curiosity, where did you get this firmware from? Looks like something has gone wrong with the version information string.


Firmware builded  25 or 26th October this script:

```
#!/bin/bash

export BUILD_DIR=$(pwd)

echo "--- CLONING MICROPYTHON ---"
git clone --depth 1 https://github.com/micropython/micropython.git
cd $BUILD_DIR/micropython
git checkout -b master
git submodule update --init --recursive

echo "--- CLONING ESP-IDF ---"
cd $BUILD_DIR/micropython/
git clone -b v5.0.2 --recursive https://github.com/espressif/esp-idf.git

echo "--- INSTALL ESP-IDF ---"
cd $BUILD_DIR/micropython/esp-idf
git checkout v5.0.2
git submodule update --init --recursive
./install.sh
. ./export.sh

echo "--- MPY-CROSS ---"
cd $BUILD_DIR/micropython/mpy-cross
LC_ALL=en_US.UTF-8 make

echo "--- ESP32 SUBMODULES ---"
cd $BUILD_DIR/micropython/ports/esp32
LC_ALL=en_US.UTF-8 make clean
LC_ALL=en_US.UTF-8 make submodules

echo "--- MAKE ---"
LC_ALL=en_US.UTF-8 make
```

> This suggests that the uart is returning None from readline(). Which it's not supposed to do because the StreamReader shouldn't have been woken up unless the uart was readable.

I used read(10) instead of readline(). It worked for now

**Reply by jimmo** [reply]
> Firmware builded 25 or 26th October this script:

Great, thanks for answering, that explains it.  (`depth=1` means it doesn't have tags, so it falls back to just the git hash, which is the expected behavior). I was just on the look-out for stuff like this because we changed how this works recently.

> I used read(10) instead of readline(). It worked for now

OK we'll need to investigate further. I'll try and repro it here.
